### PR TITLE
Gateway 171 snapshot mount test

### PR DESCRIPTION
**Pull request by rusenask**
Unmount all snapshots and fs before rollback/stash branch. 

Pipeline: https://gitlab.dotmesh.com/dotmesh/dotmesh/pipelines/20374

### CODE REVIEW

File: pkg/zfs/fs_util.go [hunk 1]
```diff
@@ -23,3 +27,47 @@ func FullIdWithSnapshot(filesystemId, snapshotId string) string {
...
+	for {
+		line, err := r.ReadString('\n')
+
+		if line != "" {
+			parts := strings.Split(line, " ")
+			if len(parts) >= 11 {
+				fsType := parts[8]
+				mountpoint := parts[4]
+				// mountedFS := parts[9]
+				// TODO: make this filter out only _our_ filesystem mounts (e.g. snapshot mounts)
```
Comment (Godley) [inline pkg/zfs/fs_util.go]: Long todo - are we doing this now or is it extra complexity that's not yet needed?

File: pkg/zfs/fs_util.go [hunk 2]
```diff
@@ -23,3 +27,47 @@ func FullIdWithSnapshot(filesystemId, snapshotId string) string {
...
+			} else {
+				return mountpoints, err
+			}
+		}
+	}
+	// reversing order so we get the snapshots first
+	for i := len(mountpoints)/2 - 1; i >= 0; i-- {
+		opp := len(mountpoints) - 1 - i
+		mountpoints[i], mountpoints[opp] = mountpoints[opp], mountpoints[i]
+	}
```
Comment (Godley) [inline pkg/zfs/fs_util.go]: This for loop confuses me a bit - so...we're only taking the first half? And we're flipping the first half with the second half? Why?

File: pkg/zfs/zfs.go [hunk 3]
```diff
@@ -1035,3 +1008,87 @@ func (z *zfs) Diff(filesystemID, snapshot, snapshotOrFilesystem string) ([]types
...
+				"namespace":  ns,
+			}).Error("failed to clear mountpoint")
+			return fmt.Errorf("failed to clear filesystem %s mountpoint %s, error: %s", filesystem, m, err)
+		}
+	}
+
+	return nil
+}
+
+func unmount(ns, mountpoint string) error {
```
Comment (Godley) [inline pkg/zfs/zfs.go]: did we not have an unmount bit of code elsewhere?

File: pkg/zfs/fs_util.go (hunk 1 above)
Comment (rusenask) [reply pkg/zfs/fs_util.go]: thanks, removed, it's already filtering :)

File: pkg/zfs/fs_util.go (hunk 2 above)
Comment (rusenask) [reply pkg/zfs/fs_util.go]: actually it just reverses the array without allocation (we need to unmount snapshots first), here's a test that shows it:

```golang
var procMountInfoTestDataWithSnapshots = `4761 22 0:167 / /var/lib/docker/containers/d3be9bb3305b151f230d3b117650686621888adf40fb008f7d78452fae9e79de/mounts/shm rw,nosuid,nodev,noexec,relatime shared:550 - tmpfs shm rw,size=65536k
206 29 0:3 net:[4026533394] /run/docker/netns/146f70b42e20 rw shared:1558 - nsfs nsfs rw
88 22 0:97 / /var/lib/docker/overlay2/ea5ed091b17dd647581bbc7543923174fe467eb26456b264211513838f7b59c5/merged rw,relatime shared:85 - overlay overlay rw,lowerdir=/var/lib/docker/overlay2/l/BZXFAXJFMGHEUOVKUCY25W27YN:/var/lib/docker/overlay2/l/YWZNJVAQQAWBNGXLWXOI5S4K4M:/var/lib/docker/overlay2/l/3GWAFBEWHPAAAOKNA3FLXYIS75:/var/lib/docker/overlay2/l/IWDTCI2G3EJ2YJNPQWIPJCSVEV:/var/lib/docker/overlay2/l/R5OT4FQ42MGCFI4DLSDXT6CIOZ:/var/lib/docker/overlay2/l/HMAMPU53HP353HISEIWQV7HCXR:/var/lib/docker/overlay2/l/I7XFZ7LAOJUV44HUEYVLWOIVYX:/var/lib/docker/overlay2/l/RTO36FZCXP2ILJBK6YAGL54DRC:/var/lib/docker/overlay2/l/SYAGX7XNICEJBOXOHSBEZX4OTR:/var/lib/docker/overlay2/l/UZGIS3UTNFEFIAYHTK7DHJZ752,upperdir=/var/lib/docker/overlay2/ea5ed091b17dd647581bbc7543923174fe467eb26456b264211513838f7b59c5/diff,workdir=/var/lib/docker/overlay2/ea5ed091b17dd647581bbc7543923174fe467eb26456b264211513838f7b59c5/work
569 22 0:234 / /var/lib/docker/containers/4b68cc32a9b195e21fdec009c4b707cc6c9a6d36f1e961ca5c5c74e6bd2b96ec/mounts/shm rw,nosuid,nodev,noexec,relatime shared:120 - tmpfs shm rw,size=65536k
913 22 0:461 / /var/lib/dotmesh/mnt/dmfs/6231f88a-630c-4d8d-8749-647020ea1391 rw,noatime shared:1566 - zfs pool/dmfs/6231f88a-630c-4d8d-8749-647020ea1391 rw,xattr,noacl
1611 22 0:565 / /var/lib/dotmesh/mnt/dmfs/6231f88a-630c-4d8d-8749-647020ea1391@a5e72133-6521-4724-825c-c7099f1012f4 ro,noatime shared:1570 - zfs pool/dmfs/6231f88a-630c-4d8d-8749-647020ea1391@a5e72133-6521-4724-825c-c7099f1012f4 ro,xattr,noacl
2042 22 0:566 / /var/lib/dotmesh/mnt/dmfs/6231f88a-630c-4d8d-8749-647020ea1391@8d3fa57d-d2c6-4886-9954-d1b504f9b171 ro,noatime shared:1863 - zfs pool/dmfs/6231f88a-630c-4d8d-8749-647020ea1391@8d3fa57d-d2c6-4886-9954-d1b504f9b171 ro,xattr,noacl
2466 22 0:639 / /var/lib/dotmesh/mnt/dmfs/6231f88a-630c-4d8d-8749-647020ea1391@5925f87f-e589-4081-861c-cced95ab8316 ro,noatime shared:1867 - zfs pool/dmfs/6231f88a-630c-4d8d-8749-647020ea1391@5925f87f-e589-4081-861c-cced95ab8316 ro,xattr,noacl
3457 22 0:640 / /var/lib/dotmesh/mnt/dmfs/6231f88a-630c-4d8d-8749-647020ea1391@5f3b3f90-50b3-441a-9e8b-b90b780514a2 ro,noatime shared:1871 - zfs pool/dmfs/6231f88a-630c-4d8d-8749-647020ea1391@5f3b3f90-50b3-441a-9e8b-b90b780514a2 ro,xattr,noacl
3521 22 0:645 / /var/lib/dotmesh/mnt/dmfs/de49b8d1-f88b-4c87-8ad4-eca8e9f157b5 rw,noatime shared:1875 - zfs pool/dmfs/de49b8d1-f88b-4c87-8ad4-eca8e9f157b5 rw,xattr,noacl
3667 22 0:646 / /var/lib/dotmesh/mnt/dmfs/de49b8d1-f88b-4c87-8ad4-eca8e9f157b5@5f3b3f90-50b3-441a-9e8b-b90b780514a2 ro,noatime shared:1879 - zfs pool/dmfs/de49b8d1-f88b-4c87-8ad4-eca8e9f157b5@5f3b3f90-50b3-441a-9e8b-b90b780514a2 ro,xattr,noacl
3767 22 0:647 / /var/lib/dotmesh/mnt/dmfs/5887919c-c980-4d5f-983d-edd0f0d76be3 rw,noatime shared:1883 - zfs pool/dmfs/5887919c-c980-4d5f-983d-edd0f0d76be3 rw,xattr,noacl
3795 22 0:648 / /var/lib/dotmesh/mnt/dmfs/5887919c-c980-4d5f-983d-edd0f0d76be3@45156382-b873-41c9-91ee-7588481452a9 ro,noatime shared:1887 - zfs pool/dmfs/5887919c-c980-4d5f-983d-edd0f0d76be3@45156382-b873-41c9-91ee-7588481452a9 ro,xattr,noacl
3808 22 0:649 / /var/lib/dotmesh/mnt/dmfs/5887919c-c980-4d5f-983d-edd0f0d76be3@2a347076-edee-4030-82b3-c22b2a6d0100 ro,noatime shared:1891 - zfs pool/dmfs/5887919c-c980-4d5f-983d-edd0f0d76be3@2a347076-edee-4030-82b3-c22b2a6d0100 ro,xattr,noacl
3835 22 0:650 / /var/lib/dotmesh/mnt/dmfs/5887919c-c980-4d5f-983d-edd0f0d76be3@1461c8df-674a-4842-a0af-e9b370299e4f ro,noatime shared:1895 - zfs pool/dmfs/5887919c-c980-4d5f-983d-edd0f0d76be3@1461c8df-674a-4842-a0af-e9b370299e4f ro,xattr,noacl`

func TestMountpointFilterWithSnapshots(t *testing.T) {
	buf := bytes.NewBufferString(procMountInfoTestDataWithSnapshots)

	reader := bufio.NewReader(buf)

	filtered, err := filterMountpoints("/var/lib/dotmesh/mnt", "5887919c-c980-4d5f-983d-edd0f0d76be3", reader)
	if err != nil {
		t.Errorf("failed to filter: %s", err)
	}
	if len(filtered) != 4 {
		t.Errorf("expected 4 entry, got: %d", len(filtered))
	}
	if filtered[0] != "/var/lib/dotmesh/mnt/dmfs/5887919c-c980-4d5f-983d-edd0f0d76be3@1461c8df-674a-4842-a0af-e9b370299e4f" {
		t.Errorf("unexpected first entry")
	}
	if filtered[3] != "/var/lib/dotmesh/mnt/dmfs/5887919c-c980-4d5f-983d-edd0f0d76be3" {
		t.Errorf("expected last entry to be without snapshot, got: %s", filtered[3])
	}
	t.Log(filtered)
}

```

File: pkg/zfs/zfs.go (hunk 3 above)
Comment (rusenask) [reply pkg/zfs/zfs.go]: this one unmounts from different namespaces :|

File: pkg/zfs/fs_util.go (hunk 2 above)
Comment (Godley) [reply pkg/zfs/fs_util.go]: Ahhhhhh ok. Fancy

**Pull request merged**